src/main/java/uk/nhs/tis/trainee/notifications/service/ProgrammeMembershipService.java

```diff
@@ -0,0 +1,73 @@
+ *
+ * Copyright 2023 Crown Copyright (Health Education England)
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and
+ * associated documentation files (the "Software"), to deal in the Software without restriction,
+ * including without limitation the rights to use, copy, modify, merge, publish, distribute,
+ * sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all copies or
+ * substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT
+ * NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+ * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package uk.nhs.tis.trainee.notifications.service;
+
+import java.util.List;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Service;
+import uk.nhs.tis.trainee.notifications.dto.Curriculum;
+import uk.nhs.tis.trainee.notifications.dto.ProgrammeMembershipEvent;
+
+/**
+ * A service for Programme memberships.
+ */
+
+@Slf4j
+@Service
+public class ProgrammeMembershipService {
+
+  private static final List<String> INCLUDE_CURRICULUM_SUBTYPES
+      = List.of("MEDICAL_CURRICULUM", "MEDICAL_SPR");
+  private static final List<String> EXCLUDE_CURRICULUM_SPECIALTIES
+      = List.of("PUBLIC HEALTH MEDICINE", "FOUNDATION");
+
+  /**
+   * Determines whether a programme membership is excluded or not, on the basis of curricula.
+   *
+   * <p>Excluded means the trainee will not be notified (contacted) in respect of this
+   * programme membership.
+   *
+   * <p>This will be TRUE if any of the following are true in relation to the curricula:
+   * 1. None have curriculumSubType = MEDICAL_CURRICULUM or MEDICAL_SPR
+   * 2. Any have specialtyName = 'Public health medicine'
+   * 3. Any have specialtyName = 'Foundation'.
+   *
+   * @param programmeMembership the Programme membership.
+   * @return true if the programme membership is excluded.
+   */
+  public boolean isExcluded(ProgrammeMembershipEvent programmeMembership) {
+    List<Curriculum> curricula = programmeMembership.curricula();
+    if (curricula == null) {
+      return true;
+    }
+
+    boolean hasMedicalSubType = curricula.stream()
+        .anyMatch(c ->
+            INCLUDE_CURRICULUM_SUBTYPES.contains(c.curriculumSubType().toUpperCase()));
+
+    boolean hasExcludedSpecialty = curricula.stream()
+        .anyMatch(c ->
+            EXCLUDE_CURRICULUM_SPECIALTIES.contains(c.curriculumSpecialty().toUpperCase()));
```

Comment:
More a stylistic thing, but I slightly prefer splitting these in to two distinct steps on the stream
```java
boolean hasMedicalSubType = curricula.stream()
     .map(c -> c.curriculumSubType().toUpperCase())
     .anyMatch(INCLUDE_CURRICULUM_SUBTYPES::contains);
```

Reply:
That does look neater.